sql/core/src/test/scala/org/apache/spark/sql/QueryTest.scala

```diff
@@ -350,6 +350,9 @@ object QueryTest {
     Row.fromSeq(row.toSeq.map {
       case null => null
       case d: java.math.BigDecimal => BigDecimal(d)
+      // Round up the values to avoid approximate error
+      case f : Float => BigDecimal(f).setScale(8, BigDecimal.RoundingMode.HALF_UP).toFloat
```

Comment:
Here is my test for your case:
scala> val a = 3.000000000000001e14
a: Double = 3.000000000000001E14

scala> val b = 3.000000000000002e14
b: Double = 3.000000000000002E14

scala> val c = BigDecimal(a).setScale(16, BigDecimal.RoundingMode.HALF_UP).toDouble
c: Double = 3.000000000000001E14

scala> val d = BigDecimal(b).setScale(16, BigDecimal.RoundingMode.HALF_UP).toDouble
d: Double = 3.000000000000002E14

scala> c == d
res3: Boolean = false

So in your case, it will return false, meaning 3.000000000000001e14 != 3.000000000000002e14, as it is expected.